### DISCUSSION TITLE
Move 3D camera forward and backward

### DISCUSSION THREAD

**Discussion by salah-walid** [iTwin/itwinjs-core#4759] [Q&A] [answered]
Hello ITwin team,

I am trying to add button on the UI to zoom in and out in the current view, using [this](https://github.com/iTwin/itwinjs-core/blob/ad74936db1b353ae2edf128b55b6696c6c371014/core/frontend/src/tools/ViewTool.ts#L1624) code as an example.

Here is my version:
```typescript
  const vp = IModelApp.viewManager.selectedView;
  if(vp?.view.is3d()) {
    const frustum = vp.getFrustum().clone();
    const oldEyePoint = vp.view.getEyePoint().clone();
    const transform = Transform.createTranslation(oldEyePoint).multiplyTransformMatrix3d(vp.view.getRotation());
    const newEyePoint = transform.multiplyXYZ(factor, 0, 0);
    const cameraOffset = Vector3d.createStartEnd(oldEyePoint, newEyePoint);
    Transform.createTranslation(cameraOffset, transform);
    frustum.transformBy(transform, frustum);
    vp.view.setupFromFrustum(frustum);
    vp.synchWithView();
  }
```
It doesn't work as intended, is there a way to solve this particular problem

**Comment by pmconne** [comment, marked as answer]
> It doesn't work

Be specific.

**Reply by salah-walid** [reply]
![app](https://user-images.githubusercontent.com/29980658/205626892-f65a55c9-464a-4564-9b15-d191ac825feb.gif)

Here is an example, it moves on a different axis it doesn't move the camera forward necessarly, it seems to move in world space instead.

**Reply by bbastings** [reply]
For 2d views or 3d views without perspective you can just use Viewport.zoom. For perspective views, it's best to use one of the "lookAt" methods. You could do something like this:

`const zoomRatio = 1.5;
const viewport = IModelApp.viewManager.selectedView;
const view = viewport.view;

if (view.is3d() && view.isCameraOn) {
  const targetPoint = view.getTargetPoint();
  const transform = Transform.createFixedPointAndMatrix(targetPoint, Matrix3d.createScale(zoomRatio, zoomRatio, zoomRatio));
  const eyePoint = view.getEyePoint().clone();
  transform.multiplyPoint3d(eyePoint, eyePoint);

  const zDir = view.getZVector();
  targetPoint.setFrom(eyePoint.plusScaled(zDir, zDir.dotProduct(eyePoint.vectorTo(targetPoint))));

  if (ViewStatus.Success === view.lookAt({eyePoint, targetPoint, upVector: view.getYVector(), lensAngle: view.camera.lens }))
    this.viewport.synchWithView();
  return;
  }

viewport.zoom(undefined, zoomRatio);`

**Reply by salah-walid** [reply]
Thank you @bbastings

**Reply by bbastings** [reply]
This [commit](https://github.com/iTwin/itwinjs-core/commit/74cea780ec39ece985baee8d5b9c4693004e80f9) adds the above logic for perspective views to Viewport.zoom. HTH
